index.js

```diff
@@ -45,7 +35,7 @@ var defer = typeof setImmediate === 'function'
 
 function onFinished (msg, listener) {
   if (isFinished(msg) !== false) {
-    defer(listener, null, msg)
+    setImmediate(listener, null, msg)
     return msg
   }
```

Comment:
The `if` block can be removed if we only accept Streams. There is a test with an unknown object blocking this change:
```js
 it('should invoke listener given an unknown object', function (done) {
    onFinished({}, done)
  })
  ```